Allow specifying both `devices` and `axis_size` to pmap.

skye opened the pull request:
This allows providing custom device assignments to nested pmaps or pmap-of-sharded_jit when running on a multi-host platform.

skye committed:
Allow specifying both `devices` and `axis_size` to pmap.

skye commented:
@trevorcai can you review? I actually ran a test script this time, I'll send you that separately :)

trevorcai commented on jax/interpreters/pxla.py:
```diff
@@ -1111,8 +1108,8 @@ def _pmap_translation_rule(c, axis_env,
                      "or another pmap is not supported.")
   # We in-line here rather than generating a Call HLO as in the xla_call
   # translation rule just because the extra tuple stuff is a pain.
-  if axis_env.devices is not None or (axis_env.names and devices is not None):
```
Missing context - why is it ok to get rid of the axis_env.devices check here?

skye replied:
axis_env.devices are devices from any outer pmaps. So that's ok now (since the outermost pmap can have devices, even if there's an inner pmap), but we still want to error if there are devices on the inner pmap (the remaining check).

trevorcai replied:
makes sense, LGTM once tests pass.

trevorcai reviewed (commented):
LGTM!

skye commented:
Ha, I tricked you, the previous version of this PR was broken! :sweat_smile: I added a unit test to exercise the non-multihost case, and verified the multi-host test script still runs. Can you take another look since I changed the logic pretty significantly?

(This suggests to me that we need a different API that's not so difficult to reason about, but I think this works for now...)

jekbradbury commented on docs/faq.rst:
```diff
@@ -177,7 +177,7 @@ concrete vs. abstract values, you may want to read `Different kinds of JAX value
 ------------------------------
 
-In the process of transforming functions, JAX replaces some some function
+In the process of transforming functions, JAX replaces some function
```
Ahaha this is the change someone made in google3 yesterday and I accidentally approved; I was supposed to make a matching PR so it wouldn't get dropped but having it brought in here through reverse-copybara is very convenient!

jekbradbury commented on jax/interpreters/pxla.py:
```diff
@@ -663,24 +663,34 @@ def parallel_callable(fun, backend, axis_name, axis_size, global_axis_size,
+    if xb.host_count() == 1:
       global_axis_size = axis_size
+    elif devices:
+      # This allows each host in a multi-host pmap to run on a different number
```
woah I had no idea that worked!

jekbradbury commented on jax/interpreters/pxla.py:
```diff
@@ -663,24 +663,34 @@ def parallel_callable(fun, backend, axis_name, axis_size, global_axis_size,
+      no_nested_sharding = True
+    else:
+      # This assumes all hosts run on the same number of devices. We make sure
+      # this assumption is true by requiring that the pmap is run on all devices
```
This is probably fine for now, but there are definitely use cases for e.g. running a pmap on half of the devices on each host where you might still want to use nesting.

skye replied:
You can do that, but you have to manually specify `axis_size` in addition to `devices` (note you only get here if global_axis_size is None and multi-host and `devices` isn't specified). There's an error message below recommending to specify `axis_size`. This is extremely difficult to reason about though, I'd like to somehow simplify it but I'm not sure how.
